File: backend-src/UZonMailDB/SQL/Core/Settings/AppSetting.cs
using Innofactor.EfCoreJsonValueConverter;
using Newtonsoft.Json.Linq;
using UZonMail.DB.SQL.Base;

namespace UZonMail.DB.SQL.Core.Settings
{
    public enum AppSettingType
    {
        System = 1,
        Organization = 2,
        User = 4
    }

    /// <summary>
    /// 系统级设置全系统只此一份
    /// 每一条记录是一个设置项
    /// </summary>
    public class AppSetting : SqlId
    {
        /// <summary>
        /// 用户 Id
        /// </summary>
        public long UserId { get; set; } = 0;

        /// <summary>
        /// 组织 Id
        /// </summary>
        public long OrganizationId { get; set; } = 0;

        /// <summary>
        /// 系统设置类型
        /// 为空是为了兼容旧数据
        /// </summary>
        public AppSettingType? Type { get; set; } = AppSettingType.System;

        /// <summary>
        /// 名称
        /// </summary>
        public string Key { get; set; }

        /// <summary>
        /// 字符串值
        /// </summary>
        public string? StringValue { get; set; }

        /// <summary>
        /// bool 值
        /// </summary>
        public bool BoolValue { get; set; }

        /// <summary>
        /// int 类型值
        /// </summary>
        public int IntValue { get; set; }

        /// <summary>
        /// Long 类型的值
        /// </summary>
        public long LongValue { get; set; }

        /// <summary>
        /// 日期时间
        /// </summary>
        public DateTime DateTime { get; set; }

        /// <summary>
        /// Json 数据
        /// </summary>
        [JsonField]
        public JToken? Json { get; set; }

        #region 静态字符串
        public static string BaseApiUrl => "baseApiUrl";
        #endregion
    }
}

File: ui-src/src/api/appSetting.ts
/* eslint-disable @typescript-eslint/no-explicit-any */


import { httpClient } from 'src/api//base/httpClient'
import { useConfig } from 'src/config'

export enum AppSettingType {
  System = 1,
  Organization = 2,
  User = 4
}

export interface IAppSetting {
  key: string,
  type: AppSettingType,
  userId: number,
  organizationId: number,
  stringValue?: string | null,
  boolValue?: boolean,
  intValue?: number,
  longValue?: number,
  dateTime?: string,
  json?: Record<string, any>
}

// #region 通用设置
/**
 * 获取用户设置
 * @returns
 */
export function updateServerBaseApiUrl () {
  const config = useConfig()
  return httpClient.put<boolean>('/app-setting/base-api-url', {
    data: {
      baseApiUrl: config.baseUrl
    }
  })
}

/**
 * 更新单个字符串型系统设置
 * @param key
 * @param value
 * @returns
 */
export function updateSystemSettingString (key: string, value: string, type: AppSettingType = AppSettingType.System) {
  return httpClient.put<boolean>('/app-setting/string', {
    params: {
      key,
      value,
      type
    }
  })
}

/**
 * 更新单个数字型系统设置
 * @param key
 * @param value
 * @returns
 */
export function updateSystemSettingBoolean (key: string, value: boolean, type: AppSettingType = AppSettingType.System) {
  return httpClient.put<boolean>('/app-setting/boolean', {
    params: {
      key,
      value,
      type
    }
  })
}

/**
 * 批量更新系统设置
 * @param key
 * @param value
 * @returns
 */
export function updateSystemSettingJson (key: string, value: Record<string, any>, type: AppSettingType = AppSettingType.System) {
  return httpClient.put<boolean>('/app-setting/json', {
    params: {
      key,
      type
    },
    data: value
  })
}

/**
 * 获取系统设置
 * @param key
 * @returns
 */
export function getSystemSetting<T = IAppSetting> (key: string, type: AppSettingType = AppSettingType.System) {
  return httpClient.get<T>('/app-setting/kv', {
    params: {
      key,
      type
    }
  })
}
// #endregion

// #region 发送设置
export interface ISendingSetting {
  userId: string,
  maxSendCountPerEmailDay: number,
  minOutboxCooldownSecond: number,
  maxOutboxCooldownSecond: number,
  maxSendingBatchSize: number,
  minInboxCooldownHours: number,
  replyToEmails?: string,
  changeIpAfterEmailCount: number,
  maxCountPerIPDomainHour: number,
}

/**
 * 获取发送设置
 * @returns
 */
export function getSendingSetting (type: AppSettingType = AppSettingType.System) {
  return httpClient.get<ISendingSetting>('/app-setting/sending-setting', {
    params: {
      type
    }
  })
}

/**
 * 更新发送设置
 * @param sendingSetting
 * @returns
 */
export function updateSendingSetting (sendingSetting: ISendingSetting, type: AppSettingType = AppSettingType.System) {
  return httpClient.put<Record<string, any>>('/app-setting/sending-setting', {
    params: {
      type
    },
    data: sendingSetting
  })
}
// #endregion
